beacon_node/beacon_chain/src/light_client_finality_update_verification.rs

```diff
@@ -0,0 +1,94 @@
+    BeaconChain, BeaconChainError, BeaconChainTypes,
+};
+use derivative::Derivative;
+use slot_clock::SlotClock;
+use std::time::Duration;
+use strum::AsRefStr;
+use types::LightClientFinalityUpdate;
+
+/// Returned when a light client finality update was not successfully verified. It might not have been verified for
+/// two reasons:
+///
+/// - The light client finality message is malformed or inappropriate for the context (indicated by all variants
+///   other than `BeaconChainError`).
+/// - The application encountered an internal error whilst attempting to determine validity
+///   (the `BeaconChainError` variant)
+#[derive(Debug, AsRefStr)]
+pub enum Error {
+    /// Light client finality update message with a lower or equal finalized_header slot already forwarded.
+    FinalityUpdateAlreadySeen,
+    /// The light client finality message was received is prior to one-third of slot duration passage. (with
+    /// respect to the gossip clock disparity and slot clock duration).
+    ///
+    /// ## Peer scoring
+    ///
+    /// Assuming the local clock is correct, the peer has sent an invalid message.
+    TooEarly,
+    /// Light client finality update message does not match the locally constructed one.
+    InvalidLightClientFinalityUpdate,
+    // Signature slot start time is none.
+    SigSlotStartIsNone,
+    // Beacon chain error occured.
+    BeaconChainError(BeaconChainError),
+}
+
+impl From<BeaconChainError> for Error {
+    fn from(e: BeaconChainError) -> Self {
+        Error::BeaconChainError(e)
+    }
+}
+
+/// Wraps a `LightClientFinalityUpdate` that has been verified for propagation on the gossip network.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "T: BeaconChainTypes"))]
+pub struct VerifiedLightClientFinalityUpdate<T: BeaconChainTypes> {
+    light_client_finality_update: LightClientFinalityUpdate<T::EthSpec>,
+    seen_timestamp: Duration,
+}
+
+impl<T: BeaconChainTypes> VerifiedLightClientFinalityUpdate<T> {
+    /// Returns `Ok(Self)` if the `light_client_finality_update` is valid to be (re)published on the gossip
+    /// network.
+    pub fn verify(
+        light_client_finality_update: LightClientFinalityUpdate<T::EthSpec>,
+        chain: &BeaconChain<T>,
+        seen_timestamp: Duration,
+    ) -> Result<Self, Error> {
+        let gossiped_finality_slot = light_client_finality_update.finalized_header.slot;
+        let local_finality_update_slot = chain.get_finalized_slot()?;
+        let one_third_slot_duration = Duration::new(chain.spec.seconds_per_slot / 3, 0);
+        let signature_slot = light_client_finality_update.signature_slot;
+
+        let start_time = chain.slot_clock.start_of(signature_slot);
+
+        // verify that no other finality_update with a lower or equal
+        // finalized_header.slot was already forwarded on the network
+        if local_finality_update_slot >= gossiped_finality_slot {
+            return Err(Error::FinalityUpdateAlreadySeen);
+        }
```

Comment:
You implemented the check correctly, but I think the spec is currently written wrong. I've opened a PR to correct it: https://github.com/ethereum/consensus-specs/pull/3086.

Apart from that, I think we need to actually store the seen finality updates so that we don't immediately discard the finality update that's equal to our view of finality. We can implement this using a new field on the `beacon_chain` which tracks the updates already seen. I _think_ we can avoid most of the complexity of `ObservedAggregates` by only storing the most recently seen finality update/optimistic update. Then our logic can be something like:

```rust
if gossiped_finality_slot <= latest_seen_update.finalized_head.slot {
    return Err(Error::FinalityUpdateAlreadySeen);
}

// .. other checks ..

*latest_seen_update = light_client_finality_update;
```